checks/run-staticAnalysis.sh

```diff
@@ -0,0 +1,135 @@
+  fileinfo=$(file -k "$file")
+  if [[ $fileinfo == *"CRLF"* ]]; then
+    echo "The following file contains CRLF line endings: $file"
+    echo "  Fix with:"
+    echo "    sed -i 's/\r$//' \"$file\""
+              sed -i 's/\r$//' "$file"
+    warnings+="At least one file contains CRLF line endings\n"
+  fi
+
+  # Check if file starts with BOM
+  if [[ $fileinfo == *"BOM"* ]]; then
+    echo "The following file starts with BOM: $file"
+    echo "  Fix with:"
+    echo "    sed -i '1s/^\xEF\xBB\xBF//' \"$file\""
+    warnings+="At least one file starts with BOM\n"
+  fi
+
+  # Check if file ends with newline
+  if [[ -n "$(tail -c 1 "$file")" ]]; then
+    echo "The following file does not end with newline: $file"
+    echo "  Fix with:"
+    echo "    sed -i -e '\$a\' \"$file\""
+    warnings+="At least one file does not end with newline\n"
+  fi
+
+done
+
+printf "\n\n\n"  # Some space to make output clearer
+
+# Only print warnings once to job summary
+warnings=$(printf "$warnings" | sort | uniq)
+warnings="# Warnings\n"$warnings"\n\n"
+
+# Run cpplint
+
+# Lines starting with one minus sign are filters which exclude certain rules
+
+#-runtime/int,\
+#-readability/alt_tokens,\
+#-whitespace/blank_line,\
+#-whitespace/braces,\
+#-whitespace/comma,\
+#-whitespace/comments,\
+#-whitespace/indent,\
+#-whitespace/operators,\
+#-whitespace/parens,\
+
+currentVersion=$(git rev-parse --abbrev-ref HEAD)
+
+for VERSION in "new" "master"
+do
+  echo "Checking ${VERSION} version"
+
+  if [[ "${VERSION}" == "master" ]]
+  then
+    git fetch &> /dev/null
+    git switch master &> /dev/null
+  else
+    git switch ${currentVersion} &> /dev/null
+  fi
+
+  cpplint --filter=-whitespace/tab --linelength=200 --counting=detailed ${codeFiles} &> $rootFolder/staticAnalysis_${VERSION}.log
+
+  grep "Category\|Total errors found" $rootFolder/staticAnalysis_${VERSION}.log > $rootFolder/staticAnalysis_${VERSION}_summary.log
+
+done
+
+git switch ${currentVersion} &> /dev/null
+
+# Use code block for monospace font in Markdown
+warnings+="\n# cpplint\n"
+warnings+=" For details run \`cpplint --filter=-whitespace/tab --linelength=200 --counting=detailed \$( find src -name \"*.h\" -or -name \"*.cpp\" -printf \"%%p \" )\`"
+warnings+=" New or fixed warnings/errors (master <-> new commit):\n"
+warnings+="\`\`\`\n"
+warnings+="master $(printf '%54s' " ") | new\n"
+
+# Delete "--suppress-common-lines" to see all errors/warnings
+warnings_cpplint=$(diff -y --suppress-common-lines $rootFolder/staticAnalysis_master_summary.log $rootFolder/staticAnalysis_new_summary.log)
+warnings+=$warnings_cpplint
+
+# Counts the categories in which new errors were introduced
+exitcode=$(printf "$warnings_cpplint" | grep "Category" | awk '{if ($11 > $5) {count++}} END {print count}')
+
+warnings+="\n\`\`\`\n"  # Close code block for monospace font
+printf "\n$warnings\n"  # Print to job output
+printf "\n$warnings\n" >> $GITHUB_STEP_SUMMARY  # Print to job summary
+
+exit $exitcode
```

Comment:
This part will annoy us a LOT in the future I fear. Currently `cpplint` takes issue with 13581 things in the master. With This PR we reduce this only to 13050. It is inevitable that new code will contain stuff `cpplint` in it's current config will not like and then it is almost impossible to find what exactly it is since we have so many violations.
I see these options:
1. Kick out `cpplint` and give up.
2. Manually fix ALL issues `cpplint` has and keep it.
3. Add `clang-format` and configure it so that cpplint likes it.
4. Replace `cpplint` with `clang-format`

Reply:
Example output from the cited script above at the time of the comment (when this passes!):

```bash
# cpplint
 For details run `cpplint --filter=-whitespace/tab --linelength=200 --counting=detailed $( find src -name "*.h" -or -name "*.cpp" -printf "%p " )` New or fixed warnings/errors (master <-> new commit):

master                                                        | new
Category 'build/include_what_you_use' errors found: 147	      |	Category 'build/include_what_you_use' errors found: 145
Category 'build/namespaces' errors found: 106		      |	Category 'build/namespaces' errors found: 6
Category 'runtime/references' errors found: 165		      |	Category 'runtime/references' errors found: 164
Category 'whitespace/comments' errors found: 2282	      |	Category 'whitespace/comments' errors found: 2272
Category 'whitespace/end_of_line' errors found: 392	      <
Category 'whitespace/ending_newline' errors found: 24	      <
Category 'whitespace/indent' errors found: 21		      |	Category 'whitespace/indent' errors found: 19
Total errors found: 13581				      |	Total errors found: 13050
```

Reply:
Due to time reasons, I will (unfortunately) decide on option 1 and just get rid of cpplint. I will also create an issue in which I mention the "need" for a proper static code analysis/formating.

Reply:
-> #276